### ISSUE TITLE
Refine type param for getObjectValue/getObjectDetails

### ISSUE THREAD

**Issue by toddbaert** [open-feature/java-sdk#50]
I believe `public <T> T getObjectValue(...` and `public <T> FlagEvaluationDetails<T> getObjectDetails(...` etc are too generic to be useful for parsing and returning an arbitrary JSON structure from a flag backend (see [here](https://github.com/open-feature/research/blob/main/prior-art/api-comparision.md#json) for notes on vendor support for JSON objects).

Because in Java, objects are not stored as dynamically-sized/structured key/value pairs (aka object literals), to instantiate a particular object from a JSON payload, some mapping between the JSON fields and class members must occur (this is basically the whole goal of [JAXB](https://javaee.github.io/jaxb-v2/) and similar annotations). I don't think we want to do this sort of deserialization in our SDK _or even in providers_. Instead, I think we should return a simple representation of a JSON Object (perhaps simply using `javax.json`) or a `Map<String, SomeCustomInterface>` or `Map<String, Object>` (looking for suggestions here). This will allow the application author to handle this payload as they want, and instantiate the objects they need, leaving the SDK out of it. We would restrict the generic T in the aforementioned methods to this type.

**Comment by justinabrahms** [comment]
Relevant: https://launchdarkly.github.io/java-server-sdk/com/launchdarkly/sdk/LDValue.html

**Comment by justinabrahms** [comment]
@toddbaert, I was chatting w/ @s-sen.  Is there a reason why something like Jackson's `ObjectMapper` can't create things from type T in the provider code? 

```java
Node y = new Node();
Node x = getObjectValue("key", y)

// -- inside getObjectValue
ObjectMapper.getValue(innerRepresentation, T.class)
```

**Comment by toddbaert** [comment]
ObjectMapper is a very specific and opinionated JSON deserialization lib that's not part of the Java runtime natively, or even conformant to the javax API - it's Jackson functionality. It just uses conventions (and **non-javax** annotations) to do the deserialization. It will transitively add `jackson-core` and `jackson-annotations` to the SDK.

This is a "nice" solution from an end-user perspective, but it's really doing the type of deserialization I mention above. I'm hesitant to add this sort of dependency to the SDK for compatibility and size reasons.

**Comment by justinabrahms** [comment]
@toddbaert I'm not suggesting we add it to the SDK. I'm suggesting that providers have a facility to accomplish the deserialization but the API still returns `T`, which is the best API for app developers.

**Comment by toddbaert** [comment]
>Sorry that I wasn't clear. I'm not suggesting we add it to the SDK. I'm suggesting that providers have a facility to accomplish the deserialization but the API still returns T, which is the best API for app developers.

I see. We'd be making it the provider's responsibility to deserialize and instantiate these objects. Is it fair to expect providers to do this in a consistent way? It sort of forces the same compatibility and size concerns I mentioned above onto providers.

**Comment by toddbaert** [comment]
To be clear - I'm not 100% opposed to this idea - but I'd like to be really thoughtful about it when there's other options.

**Comment by justinabrahms** [comment]
It seems like returning T is the best end user experience. Open to hearing from provider implementers.

**Comment by s-sen** [comment]
Let us walk through a scenario where T is generic and specific to the provider. The provider is responsible for defining what T is. For example provider X may say T is a JSON object and as long as the user specifies a valid JSON, provider X will store it and deserialize it in the most optimal way using the most optimal lib - Jackson/Gson/Provider-X-Home-Grown etc. 
Provider Y may say that they do not support JSON object and hence when the user switches from provider X to Y,  user code needs to be changed.

Another scenario is we change the spec so that the 'structure' type is a JSONValue type. Each provider is responsible for storing it as a valid JSON entity and also  responsible for optimal deserialization of it. But the user can store any valid JSON structure there and use the JSONValue<T> signature (in java). So when the user switches provider, the user code does not need to change as long as the new provider returns the same object type after deserializing the same JSON structure.

If we specify that the type is a generic structure and platform/provider specific, then IMO there may not be a standard way to ensure that user does not get affected if a decision is made to switch the provider.

**Comment by toddbaert** [comment]
> Another scenario is we change the spec so that the 'structure' type is a JSONValue type.

I don't _think_ we need to do this, because we have it already. This is basically what the spec already says (without being too prescriptive): https://github.com/open-feature/spec/blob/main/specification/types.md#structure

> But the user can store any valid JSON structure there and use the JSONValue<T> signature (in java). So when the user switches provider, the user code does not need to change as long as the new provider returns the same object type after deserializing the same JSON structure.

>If we specify that the type is a generic structure and platform/provider specific, then IMO there may not be a standard way to ensure that user does not get affected if a decision is made to switch the provider.

I think this is basically what I'm in favor of... we can use some fairly general JSON representation, such as [JSONObject](https://docs.oracle.com/javaee/7/api/javax/json/JsonObject.html). I hope I've understood your point!

**Comment by s-sen** [comment]
Yes, we can specify a generic representation of JSON  - but  what should be that interface specification be? Is it tied to the platform/language e.g. for java does the provider need to use [JSONObject](https://docs.oracle.com/javaee/7/api/javax/json/JsonObject.html)?
 Or does the spec specify an interface and the provider needs to implement it based on the supported language?

**Comment by toddbaert** [comment]
> for java does the provider need to use [JSONObject](https://docs.oracle.com/javaee/7/api/javax/json/JsonObject.html)?

I think that's really up to us. The spec stays fairly vague, on purpose, so that languages can choose how to deal with these issues as they need. This entire problem doesn't exist in Typescript/Javascript for example, we can just return the JSON from the backend.

> Or does the spec specify an interface and the provider needs to implement it based on the supported language?

Nope, the spec gives a lot of flexibility - it just says this method should return "structured data", so a JSON structure representation is a fine choice, I think.

**Comment by justinabrahms** [comment]
I think we should also think about whether taking in `T` for eval context is also going to be a problem. I don't love them having the same name of `structure` and having different representations.

**Comment by toddbaert** [comment]
> I think we should also think about whether taking in T for eval context is also going to be a problem. I don't love them having the same name of structure and having different representations.

I sort-of feel the same way (we have the same issue in the Node SDK). I think absent some reason they can't be the same structure, they should be. One possible counterpoint to consider is that they are used differently, at least from the application author perspective: the object evaluation result is mostly read (with the exception of providers), while the context is mostly written.

@justinabrahms @s-sen do you guys have any prescription based on our discussion? Is JSONObject reasonable? What about just a Map<string, object>?

**Comment by justinabrahms** [comment]
I'd check whether JSONobject is accessible without using a special java EE version. I'm a bit lost in the spec (on mobile) to know if it's in the base java implementation. If so, let's use that.

**Comment by toddbaert** [comment]
`javax.json` is an extension, not in the SE (and it's actually called - `jakarta.json` now since the EE rebrand) but it's just an API - an implementation has to be used [org.glassfish/jakarta.json](https://mvnrepository.com/artifact/org.glassfish/jakarta.json) for example... so to use it we'd also be "making a choice", but a more flexibile one, with consumers being able to specify their implementation.

I'm still not 100% sure is the best option over something even more generic.

**Comment by justinabrahms** [comment]
Or us just defining our own interface a la LDValue

**Comment by toddbaert** [comment]
>Or us just defining our own interface a la LDValue

Absolutely.

I actually really liked this [quick write-up](https://dzone.com/articles/which-is-the-right-java-abstraction-for-json). I think we are even more restricted though because any choices we make are being forced on our consumers.

**Comment by toddbaert** [comment]
@justinabrahms @s-sen What if we extracted the relevant properties/methods from `EvaluationContext.java` ... so we'd have things like `getAttributesByType/addStringAttribute` etc (but nothing relating to `targeting key`, since that's not relevant to object return types...

Then the provider would construct an instance this object manually...

I guess the big question at the end of the day, is: is the effort of providers having to construct this object worth the benefit of avoiding the need for a JSON deserialization library such as ObjectMapper. I honestly don't know.

**Comment by toddbaert** [comment]
Another data point: Harness uses gson JsonObjects: https://github.com/harness/ff-java-server-sdk/blob/main/src/main/java/io/harness/cf/client/api/Evaluation.java

**Comment by justinabrahms** [comment]
Here's where my mind is at:

We don't really want to force a serialization choice onto providers, because it will/may conflict with the serialization choice used in the app themselves. This results in classpath issues, which is a big user pain. Instead, we'll adopt a "work it out with your provider implementation" stance, and present a generic format for SDK->provider interchange.

To that end, we'll support generating `Map<String, Object>` with a constrained set of choices for what `Object` can be. If java supported it, we would have something like:

```
JSONValue = Integer|String|Boolean
Map<String, JSONValue | List<JSONValue> | Map<JSONValue>>
```

We lose type safety, but given it's coming from the user typing things in a UI probably.. type safety was already out the window.

We already make a distinction between a `Structure` and an `Object`, which I like. `Object` comes from the server. `Structure` lives in the context.

**Comment by toddbaert** [comment]
@justinabrahms OK - I think I have something like this partially implemented. I will open a draft PR for us to build from soon.

**Comment by s-sen** [comment]
The current interface in the java SDK is this: `<T> T getObjectValue(String key, T defaultValue, EvaluationContext ctx, FlagEvaluationOptions options);`

A provider can document that according to the spec, the customer has to provide a valid JSON string value  and the provider is responsible for deserializing it efficiently and return it as `<T>`.  Example: the provider can use gson to translate `{"hello":"world"}` to a `Map<String,String>` if that is the type that the customer provides as  the type of the default value. Essentially the provider will deserialize based on the type of `<T>` passed in by the customer for the default value.  The customer can provide a deeply nested JSON structure according to their need, and as long as the default value has the correct type of <T> the java provider can deserialize it. 

The above approach is flexible and provides a lot of leeway to the customer without adding too much work on the part of the provider. But the -ve side of the above approach is that if the customer passes null as the default value then the provider cannot do anything.

`javax.jsonJSonValue`is java EE. So, it should not be added IMO in the SDK. The other option is use a `Map<String, Object>`. But this means that the customer can provide a deeply nested JSON structure only if it is responsible for deserializing it.  If we use a specific class like LDValue then each provider has to provide extensive support for this value similar to what LDValue provides. We can provide a simpler approach  e.g. `JSONValue getObjectValue(String key, JSONValue defaultValue, EvaluationContext ctx, FlagEvaluationOptions options)` where JSONValue can return a list/map etc. But not sure how this is better than a generic `Map<String, Object>`.

**Comment by justinabrahms** [comment]
The benefit of `LDValue` is that it's more type safe. It wraps integers, arrays, etc, so you know folks aren't shoving things in there that aren't of known types.

I think we may need to look at some actual code and/or hop on a call to just synchronously chat this out.

**Comment by toddbaert** [comment]
> The benefit of `LDValue` is that it's more type safe. It wraps integers, arrays, etc, so you know folks aren't shoving things in there that aren't of known types.
> 
> I think we may need to look at some actual code and/or hop on a call to just synchronously chat this out.

I'm close to having a my own PoC PR. I will update here when I have it, then maybe it will give us something practical to discuss. I got pulled into another task but I should have something tonight or tomorrow.

**Comment by s-sen** [comment]
Here is some pseudo code that uses the java SDK  interface in it's current form.

--- application code ---
// Flag value type is JSON and the application code provides the value as a JSON map and creates default value and calls the API to get the actual flag value
```
Map<String, Object> defaultMapValue = new HashMap<>();
defaultMapValue.put("color", "red");
defaultMapValue.put("height", 4);
defaultMapValue.put("enabled", false);  

Map<String,Object> flagValue = client.getObjectValue("mykey", defaultMapValue);
```
----  Provider code  (this provider uses Gson) ----
// Get the string representation of the value from the db/memory
`String jsonFlagValue = < get value for key "mykey" JSON >`

// Now use the following method to generate the value with the correct type
```
<T> deserializeJsonValue(String jsonFlagValue, T defaultValue) {

    // validate defaultValue is not null....etc.
    
    Gson gson = new Gson();
    T flagValue = gson.fromJson(jsonFlagValue, defaultValue.getClass());
  
    return flagValue;
}
```

**Comment by toddbaert** [comment]
@justinabrahms @s-sen take a look at: https://github.com/open-feature/java-sdk/pull/51

**Comment by justinabrahms** [comment]
@toddbaert @s-sen check out https://github.com/open-feature/java-sdk-contrib/pull/67 (not actually tested w/ flagd) which is the implementation using gson. It makes me a little grumpy that I'm taking the protobuf (todd's thing uses gRPC) and converting it into json so we can deserialize it with gson. I do like that users are returned a T tho. That still seems like the ideal interface to me.

**Comment by toddbaert** [comment]
@justinabrahms What you're doing there makes sense, and I'm sure in most cases a provider could produce a `T` instance if we really wanted them to. I just worry that it won't be consistently implemented, that magically instantiating a T with the correct fields set will be error prone, and that it forces dependencies on providers (gson in this case).

**Comment by rgrassian-split** [comment]
From my experience trying to implement the `T getObjectValue ` for the Split provider we ran into the issue that due to type erasure there was no way to get the type of the generic at runtime which has caused big headaches for us. I am very pro switching this

**Comment by justinabrahms** [comment]
Looks like LaunchDarkly's setup is they build this `Structure` equivalent (LDValue) and then provide adaptors for folks who want to use it w/ Jackson or Gson. ([gson example](https://launchdarkly.github.io/java-server-sdk/com/launchdarkly/sdk/json/LDGson.html))

**Comment by justinabrahms** [comment]
Closing this, as I think we solved it with #51 (and subsequent commits)
